File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Symbols/ErrorTypeSymbol.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading;
using Microsoft.Cci;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
{
	internal class ErrorTypeSymbol : NamedTypeSymbol, IErrorTypeSymbol
	{
		internal static readonly ErrorTypeSymbol UnknownResultType = new ErrorTypeSymbol();

		internal virtual DiagnosticInfo ErrorInfo => null;

		public override bool IsReferenceType => true;

		public override bool IsValueType => false;

		public override IEnumerable<string> MemberNames => SpecializedCollections.EmptyEnumerable<string>();

		public sealed override SymbolKind Kind => SymbolKind.ErrorType;

		public sealed override TypeKind TypeKind => TypeKind.Error;

		internal sealed override bool IsInterface => false;

		public override Symbol ContainingSymbol => null;

		public override ImmutableArray<Location> Locations => ImmutableArray<Location>.Empty;

		public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => ImmutableArray<SyntaxReference>.Empty;

		public override int Arity => 0;

		public override string Name => string.Empty;

		internal override bool MangleName => false;

		internal sealed override bool HasSpecialName => false;

		public sealed override bool IsSerializable => false;

		internal override TypeLayout Layout => default(TypeLayout);

		internal override CharSet MarshallingCharSet => base.DefaultMarshallingCharSet;

		internal override ImmutableArray<TypeSymbol> TypeArgumentsNoUseSiteDiagnostics => ImmutableArray<TypeSymbol>.Empty;

		internal override bool HasTypeArgumentsCustomModifiers => false;

		public override ImmutableArray<TypeParameterSymbol> TypeParameters => ImmutableArray<TypeParameterSymbol>.Empty;

		public override NamedTypeSymbol ConstructedFrom => this;

		public sealed override Accessibility DeclaredAccessibility => Accessibility.Public;

		public sealed override bool IsMustInherit => false;

		public sealed override bool IsNotInheritable => false;

		public sealed override bool MightContainExtensionMethods => false;

		internal override bool HasCodeAnalysisEmbeddedAttribute => false;

		internal override bool HasVisualBasicEmbeddedAttribute => false;

		internal override bool IsExtensibleInterfaceNoUseSiteDiagnostics => false;

		internal override bool IsWindowsRuntimeImport => false;

		internal override bool ShouldAddWinRTMembers => false;

		internal override bool IsComImport => false;

		internal override TypeSymbol CoClassType => null;

		internal override bool HasDeclarativeSecurity => false;

		internal override TypeSubstitution TypeSubstitution => null;

		internal override bool CanConstruct => false;

		internal override string DefaultPropertyName => null;

		internal NamedTypeSymbol NonErrorGuessType
		{
			get
			{
				ImmutableArray<Symbol> candidateSymbols = CandidateSymbols;
				if (candidateSymbols.Length == 1)
				{
					return candidateSymbols[0] as NamedTypeSymbol;
				}
				return null;
			}
		}

		internal virtual LookupResultKind ResultKind => LookupResultKind.Empty;

		internal sealed override ObsoleteAttributeData ObsoleteAttributeData => null;

		public virtual ImmutableArray<Symbol> CandidateSymbols => ImmutableArray<Symbol>.Empty;

		public CandidateReason CandidateReason
		{
			get
			{
				if (CandidateSymbols.IsEmpty)
				{
					return CandidateReason.None;
				}
				return LookupResultKindExtensions.ToCandidateReason(ResultKind);
			}
		}

		public ImmutableArray<ISymbol> IErrorTypeSymbol_CandidateSymbols => StaticCast<ISymbol>.From(CandidateSymbols);

		internal override UseSiteInfo<AssemblySymbol> GetUseSiteInfo()
		{
			if (base.IsDefinition)
			{
				return new UseSiteInfo<AssemblySymbol>(ErrorInfo);
			}
			return base.GetUseSiteInfo();
		}

		internal override DiagnosticInfo GetUnificationUseSiteDiagnosticRecursive(Symbol owner, ref HashSet<TypeSymbol> checkedTypes)
		{
			return null;
		}

		internal override NamedTypeSymbol MakeDeclaredBase(BasesBeingResolved basesBeingResolved, BindingDiagnosticBag diagnostics)
		{
			return null;
		}

		internal override ImmutableArray<NamedTypeSymbol> MakeDeclaredInterfaces(BasesBeingResolved basesBeingResolved, BindingDiagnosticBag diagnostics)
		{
			return ImmutableArray<NamedTypeSymbol>.Empty;
		}

		internal override NamedTypeSymbol MakeAcyclicBaseType(BindingDiagnosticBag diagnostics)
		{
			return null;
		}

		internal override ImmutableArray<NamedTypeSymbol> MakeAcyclicInterfaces(BindingDiagnosticBag diagnostics)
		{
			return ImmutableArray<NamedTypeSymbol>.Empty;
		}

		public override ImmutableArray<Symbol> GetMembers()
		{
			return ImmutableArray<Symbol>.Empty;
		}

		public override ImmutableArray<Symbol> GetMembers(string name)
		{
			return ImmutableArray<Symbol>.Empty;
		}

		public override ImmutableArray<NamedTypeSymbol> GetTypeMembers()
		{
			return ImmutableArray<NamedTypeSymbol>.Empty;
		}

		public override ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name)
		{
			return ImmutableArray<NamedTypeSymbol>.Empty;
		}

		public override ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name, int arity)
		{
			return ImmutableArray<NamedTypeSymbol>.Empty;
		}

		internal override IEnumerable<FieldSymbol> GetFieldsToEmit()
		{
			return SpecializedCollections.EmptyEnumerable<FieldSymbol>();
		}

		public override ImmutableArray<CustomModifier> GetTypeArgumentCustomModifiers(int ordinal)
		{
			return GetEmptyTypeArgumentCustomModifiers(ordinal);
		}

		internal override TResult Accept<TArgument, TResult>(VisualBasicSymbolVisitor<TArgument, TResult> visitor, TArgument arg)
		{
			return visitor.VisitErrorType(this, arg);
		}

		internal ErrorTypeSymbol()
		{
		}

		internal override ImmutableArray<string> GetAppliedConditionalSymbols()
		{
			return ImmutableArray<string>.Empty;
		}

		internal override AttributeUsageInfo GetAttributeUsageInfo()
		{
			throw ExceptionUtilities.Unreachable;
		}

		internal override IEnumerable<SecurityAttribute> GetSecurityInformation()
		{
			throw ExceptionUtilities.Unreachable;
		}

		internal override TypeWithModifiers InternalSubstituteTypeParameters(TypeSubstitution substitution)
		{
			return new TypeWithModifiers(this);
		}

		public override NamedTypeSymbol Construct(ImmutableArray<TypeSymbol> typeArguments)
		{
			throw new InvalidOperationException();
		}

		public override bool Equals(TypeSymbol obj, TypeCompareKind comparison)
		{
			return (object)this == obj;
		}

		public override int GetHashCode()
		{
			return RuntimeHelpers.GetHashCode(this);
		}

		internal override void GenerateDeclarationErrors(CancellationToken cancellationToken)
		{
			throw ExceptionUtilities.Unreachable;
		}

		internal sealed override IEnumerable<PropertySymbol> GetSynthesizedWithEventsOverrides()
		{
			return SpecializedCollections.EmptyEnumerable<PropertySymbol>();
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Symbols/ErrorTypeSymbol.vb
Imports Microsoft.Cci
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Roslyn.Utilities
Imports System
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices
Imports System.Threading

Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
	Friend Class ErrorTypeSymbol
		Inherits NamedTypeSymbol
		Implements IErrorTypeSymbol
		Friend ReadOnly Shared UnknownResultType As ErrorTypeSymbol

		Public Overrides ReadOnly Property Arity As Integer
			Get
				Return 0
			End Get
		End Property

		Friend Overrides ReadOnly Property CanConstruct As Boolean
			Get
				Return False
			End Get
		End Property

		Public ReadOnly Property CandidateReason As Microsoft.CodeAnalysis.CandidateReason Implements IErrorTypeSymbol.CandidateReason
			Get
				Dim candidateReason1 As Microsoft.CodeAnalysis.CandidateReason
				candidateReason1 = If(Not Me.CandidateSymbols.IsEmpty, Me.ResultKind.ToCandidateReason(), Microsoft.CodeAnalysis.CandidateReason.None)
				Return candidateReason1
			End Get
		End Property

		Public Overridable ReadOnly Property CandidateSymbols As ImmutableArray(Of Symbol)
			Get
				Return ImmutableArray(Of Symbol).Empty
			End Get
		End Property

		Friend Overrides ReadOnly Property CoClassType As TypeSymbol
			Get
				Return Nothing
			End Get
		End Property

		Public Overrides ReadOnly Property ConstructedFrom As NamedTypeSymbol
			Get
				Return Me
			End Get
		End Property

		Public Overrides ReadOnly Property ContainingSymbol As Symbol
			Get
				Return Nothing
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property DeclaredAccessibility As Accessibility
			Get
				Return Accessibility.[Public]
			End Get
		End Property

		Public Overrides ReadOnly Property DeclaringSyntaxReferences As ImmutableArray(Of SyntaxReference)
			Get
				Return ImmutableArray(Of SyntaxReference).Empty
			End Get
		End Property

		Friend Overrides ReadOnly Property DefaultPropertyName As String
			Get
				Return Nothing
			End Get
		End Property

		Friend Overridable ReadOnly Property ErrorInfo As DiagnosticInfo
			Get
				Return Nothing
			End Get
		End Property

		Friend Overrides ReadOnly Property HasCodeAnalysisEmbeddedAttribute As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property HasDeclarativeSecurity As Boolean
			Get
				Return False
			End Get
		End Property

		Friend NotOverridable Overrides ReadOnly Property HasSpecialName As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property HasTypeArgumentsCustomModifiers As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property HasVisualBasicEmbeddedAttribute As Boolean
			Get
				Return False
			End Get
		End Property

		Public ReadOnly Property IErrorTypeSymbol_CandidateSymbols As ImmutableArray(Of ISymbol) Implements IErrorTypeSymbol.CandidateSymbols
			Get
				Return StaticCast(Of ISymbol).From(Of Symbol)(Me.CandidateSymbols)
			End Get
		End Property

		Friend Overrides ReadOnly Property IsComImport As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property IsExtensibleInterfaceNoUseSiteDiagnostics As Boolean
			Get
				Return False
			End Get
		End Property

		Friend NotOverridable Overrides ReadOnly Property IsInterface As Boolean
			Get
				Return False
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property IsMustInherit As Boolean
			Get
				Return False
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property IsNotInheritable As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property IsReferenceType As Boolean
			Get
				Return True
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property IsSerializable As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property IsValueType As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property IsWindowsRuntimeImport As Boolean
			Get
				Return False
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property Kind As SymbolKind
			Get
				Return SymbolKind.ErrorType
			End Get
		End Property

		Friend Overrides ReadOnly Property Layout As TypeLayout
			Get
				Return New TypeLayout()
			End Get
		End Property

		Public Overrides ReadOnly Property Locations As ImmutableArray(Of Location)
			Get
				Return ImmutableArray(Of Location).Empty
			End Get
		End Property

		Friend Overrides ReadOnly Property MangleName As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property MarshallingCharSet As CharSet
			Get
				Return MyBase.DefaultMarshallingCharSet
			End Get
		End Property

		Public Overrides ReadOnly Property MemberNames As IEnumerable(Of String)
			Get
				Return SpecializedCollections.EmptyEnumerable(Of String)()
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property MightContainExtensionMethods As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property Name As String
			Get
				Return [String].Empty
			End Get
		End Property

		Friend ReadOnly Property NonErrorGuessType As NamedTypeSymbol
			Get
				Dim item As NamedTypeSymbol
				Dim candidateSymbols As ImmutableArray(Of Symbol) = Me.CandidateSymbols
				If (candidateSymbols.Length <> 1) Then
					item = Nothing
				Else
					item = TryCast(candidateSymbols(0), NamedTypeSymbol)
				End If
				Return item
			End Get
		End Property

		Friend NotOverridable Overrides ReadOnly Property ObsoleteAttributeData As Microsoft.CodeAnalysis.ObsoleteAttributeData
			Get
				Return Nothing
			End Get
		End Property

		Friend Overridable ReadOnly Property ResultKind As LookupResultKind
			Get
				Return LookupResultKind.Empty
			End Get
		End Property

		Friend Overrides ReadOnly Property ShouldAddWinRTMembers As Boolean
			Get
				Return False
			End Get
		End Property

		Friend Overrides ReadOnly Property TypeArgumentsNoUseSiteDiagnostics As ImmutableArray(Of TypeSymbol)
			Get
				Return ImmutableArray(Of TypeSymbol).Empty
			End Get
		End Property

		Public NotOverridable Overrides ReadOnly Property TypeKind As Microsoft.CodeAnalysis.TypeKind
			Get
				Return Microsoft.CodeAnalysis.TypeKind.[Error]
			End Get
		End Property

		Public Overrides ReadOnly Property TypeParameters As ImmutableArray(Of TypeParameterSymbol)
			Get
				Return ImmutableArray(Of TypeParameterSymbol).Empty
			End Get
		End Property

		Friend Overrides ReadOnly Property TypeSubstitution As Microsoft.CodeAnalysis.VisualBasic.Symbols.TypeSubstitution
			Get
				Return Nothing
			End Get
		End Property

		Shared Sub New()
			ErrorTypeSymbol.UnknownResultType = New ErrorTypeSymbol()
		End Sub

		Friend Sub New()
			MyBase.New()
		End Sub

		Friend Overrides Function Accept(Of TArgument, TResult)(ByVal visitor As VisualBasicSymbolVisitor(Of TArgument, TResult), ByVal arg As TArgument) As TResult
			Return visitor.VisitErrorType(Me, arg)
		End Function

		Public Overrides Function Construct(ByVal typeArguments As ImmutableArray(Of TypeSymbol)) As NamedTypeSymbol
			Throw New InvalidOperationException()
		End Function

		Public Overrides Function Equals(ByVal obj As TypeSymbol, ByVal comparison As TypeCompareKind) As Boolean
			Return Me = obj
		End Function

		Friend Overrides Sub GenerateDeclarationErrors(ByVal cancellationToken As System.Threading.CancellationToken)
			Throw ExceptionUtilities.Unreachable
		End Sub

		Friend Overrides Function GetAppliedConditionalSymbols() As ImmutableArray(Of String)
			Return ImmutableArray(Of String).Empty
		End Function

		Friend Overrides Function GetAttributeUsageInfo() As AttributeUsageInfo
			Throw ExceptionUtilities.Unreachable
		End Function

		Friend Overrides Function GetFieldsToEmit() As IEnumerable(Of FieldSymbol)
			Return SpecializedCollections.EmptyEnumerable(Of FieldSymbol)()
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return RuntimeHelpers.GetHashCode(Me)
		End Function

		Public Overrides Function GetMembers() As ImmutableArray(Of Symbol)
			Return ImmutableArray(Of Symbol).Empty
		End Function

		Public Overrides Function GetMembers(ByVal name As String) As ImmutableArray(Of Symbol)
			Return ImmutableArray(Of Symbol).Empty
		End Function

		Friend Overrides Function GetSecurityInformation() As IEnumerable(Of SecurityAttribute)
			Throw ExceptionUtilities.Unreachable
		End Function

		Friend NotOverridable Overrides Function GetSynthesizedWithEventsOverrides() As IEnumerable(Of PropertySymbol)
			Return SpecializedCollections.EmptyEnumerable(Of PropertySymbol)()
		End Function

		Public Overrides Function GetTypeArgumentCustomModifiers(ByVal ordinal As Integer) As ImmutableArray(Of CustomModifier)
			Return MyBase.GetEmptyTypeArgumentCustomModifiers(ordinal)
		End Function

		Public Overrides Function GetTypeMembers() As ImmutableArray(Of NamedTypeSymbol)
			Return ImmutableArray(Of NamedTypeSymbol).Empty
		End Function

		Public Overrides Function GetTypeMembers(ByVal name As String) As ImmutableArray(Of NamedTypeSymbol)
			Return ImmutableArray(Of NamedTypeSymbol).Empty
		End Function

		Public Overrides Function GetTypeMembers(ByVal name As String, ByVal arity As Integer) As ImmutableArray(Of NamedTypeSymbol)
			Return ImmutableArray(Of NamedTypeSymbol).Empty
		End Function

		Friend Overrides Function GetUnificationUseSiteDiagnosticRecursive(ByVal owner As Symbol, ByRef checkedTypes As HashSet(Of TypeSymbol)) As DiagnosticInfo
			Return Nothing
		End Function

		Friend Overrides Function GetUseSiteInfo() As UseSiteInfo(Of AssemblySymbol)
			Dim useSiteInfo As UseSiteInfo(Of AssemblySymbol)
			useSiteInfo = If(Not MyBase.IsDefinition, MyBase.GetUseSiteInfo(), New UseSiteInfo(Of AssemblySymbol)(Me.ErrorInfo))
			Return useSiteInfo
		End Function

		Friend Overrides Function InternalSubstituteTypeParameters(ByVal substitution As Microsoft.CodeAnalysis.VisualBasic.Symbols.TypeSubstitution) As TypeWithModifiers
			Return New TypeWithModifiers(Me)
		End Function

		Friend Overrides Function MakeAcyclicBaseType(ByVal diagnostics As Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag) As NamedTypeSymbol
			Return Nothing
		End Function

		Friend Overrides Function MakeAcyclicInterfaces(ByVal diagnostics As Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag) As ImmutableArray(Of NamedTypeSymbol)
			Return ImmutableArray(Of NamedTypeSymbol).Empty
		End Function

		Friend Overrides Function MakeDeclaredBase(ByVal basesBeingResolved As Microsoft.CodeAnalysis.VisualBasic.BasesBeingResolved, ByVal diagnostics As Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag) As NamedTypeSymbol
			Return Nothing
		End Function

		Friend Overrides Function MakeDeclaredInterfaces(ByVal basesBeingResolved As Microsoft.CodeAnalysis.VisualBasic.BasesBeingResolved, ByVal diagnostics As Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag) As ImmutableArray(Of NamedTypeSymbol)
			Return ImmutableArray(Of NamedTypeSymbol).Empty
		End Function
	End Class
End Namespace